### ISSUE TITLE
could not broadcast input array from shape (3,2048) into shape (6144)

### ISSUE THREAD

**Issue by leon3428** [plaidml/plaidml#149]
I just installed plaidml and i tried to run this example:
```python
#!/usr/bin/env python

import plaidml.keras
plaidml.keras.install_backend() 

import numpy as np
import matplotlib.pyplot as plt
from keras.models import Sequential
from keras.layers.core import Dense, Activation, Dropout
from keras.datasets import mnist
from keras.utils import np_utils

# fix a random seed for reproducibility
np.random.seed(9)

# user inputs
nb_epoch = 25
num_classes = 10
batch_size = 128
train_size = 60000
test_size = 10000
v_length = 784

# split the mnist data into train and test
(trainData, trainLabels), (testData, testLabels) = mnist.load_data()


# reshape the dataset
trainData = trainData.reshape(train_size, v_length)
testData = testData.reshape(test_size, v_length)
trainData = trainData.astype("float32")
testData = testData.astype("float32")
trainData /= 255
testData /= 255


# convert class vectors to binary class matrices --> one-hot encoding
mTrainLabels = np_utils.to_categorical(trainLabels, num_classes)
mTestLabels = np_utils.to_categorical(testLabels, num_classes)

# create the model
model = Sequential()
model.add(Dense(512, input_shape=(784,)))
model.add(Activation("relu"))
model.add(Dropout(0.2))
model.add(Dense(256))
model.add(Activation("relu"))
model.add(Dropout(0.2))
model.add(Dense(num_classes))
model.add(Activation("softmax"))

# summarize the model
model.summary()

# compile the model
model.compile(loss="categorical_crossentropy",
			  optimizer="adam",
			  metrics=["accuracy"])

# fit the model
history = model.fit(trainData, 
				 	mTrainLabels,
					validation_data=(testData, mTestLabels),
					batch_size=batch_size,
					nb_epoch=nb_epoch,
					verbose=2)

# print the history keys


# evaluate the model
scores = model.evaluate(testData, mTestLabels, verbose=0)

# history plot for accuracy
plt.plot(history.history["acc"])
plt.plot(history.history["val_acc"])
plt.title("Model Accuracy")
plt.xlabel("Epoch")
plt.ylabel("Accuracy")
plt.legend(["train", "test"], loc="upper left")
plt.show()

# history plot for accuracy
plt.plot(history.history["loss"])
plt.plot(history.history["val_loss"])
plt.title("Model Loss")
plt.xlabel("Epoch")
plt.ylabel("Loss")
plt.legend(["train", "test"], loc="upper left")
plt.show()

```
and I got this error

> could not broadcast input array from shape (3,2048) into shape (6144)

Then I tried running Hello VGG example from plaidml github page and I got the same error.

I am using plaidml 0.3.4 on ubuntu in virtualenv and I am trying to run this code on rx 480.

Tnx for help.

**Comment by brianretford** [comment]
Tim, can you take a look at this when you get a chance?

**Comment by mzeug** [comment]
Uhh got the same error. 

`Traceback (most recent call last):
  File "test.py", line 17, in <module>
    model = kapp.VGG19()
  File "/home/user/src/machine_learning/venv/lib/python3.6/site-packages/keras/applications/vgg19.py", line 112, in VGG19
    x = Conv2D(64, (3, 3), activation='relu', padding='same', name='block1_conv1')(img_input)
  File "/home/user/src/machine_learning/venv/lib/python3.6/site-packages/keras/engine/topology.py", line 575, in __call__
    self.build(input_shapes[0])
  File "/home/user/src/machine_learning/venv/lib/python3.6/site-packages/keras/layers/convolutional.py", line 134, in build
    constraint=self.kernel_constraint)
  File "/home/user/src/machine_learning/venv/lib/python3.6/site-packages/keras/legacy/interfaces.py", line 87, in wrapper
    return func(*args, **kwargs)
  File "/home/user/src/machine_learning/venv/lib/python3.6/site-packages/keras/engine/topology.py", line 396, in add_weight
    weight = K.variable(initializer(shape),
  File "/home/user/src/machine_learning/venv/lib/python3.6/site-packages/keras/initializers.py", line 212, in __call__
    dtype=dtype, seed=self.seed)
  File "/home/user/src/machine_learning/venv/lib/python3.6/site-packages/plaidml/keras/backend.py", line 1238, in random_uniform
    rng_state = _make_rng_state(seed)
  File "/home/user/src/machine_learning/venv/lib/python3.6/site-packages/plaidml/keras/backend.py", line 170, in _make_rng_state
    rng_state = variable(rng_init, dtype='uint32')
  File "/home/user/src/machine_learning/venv/lib/python3.6/site-packages/plaidml/keras/backend.py", line 1762, in variable
    view.copy_from_ndarray(value)
  File "/home/user/src/machine_learning/venv/lib/python3.6/site-packages/plaidml/__init__.py", line 1197, in copy_from_ndarray
    np.copyto(dst, src)
`

**Comment by Gizmomens** [comment]
Same issue, I'm using Keras Version 2.2.0 and Plaidml backend version0.3.4

**Comment by mzeug** [comment]
Took a realy quick look and toyed with the code.
Seems that the two methods "copy_from_ndarray" and "copy_to_ndarray" use the function np.ctypeslib.as_array(self, shape=src.shape). "self" is from type "class _view" and therefore i asume ctypeslib.as_array, according to the documentation, ignores the shape parameter. 
Intuitively i tried np.ctypeslib.as_array(self._base, shape=src.shape) in both methods and this seems to work.
Im really new to this topic so maybe (propably!) this is complet nonsense but i haven't had the time yet to took a deeper look.

**Comment by arytbk** [comment]
same here, "plaidbench keras mobilenet" will give "could not broadcast input array from shape (3,2048) into shape (6144)".

**Comment by dr-romster** [comment]
I am seeing the same line errors as @mzeug from a new install as well when trying `plaidbench keras mobilenet`, or upon when attempting `model.add( )` as per @leon3428 's example.

I have tried each of the available 4 options in the `plaidml-setup` 
[ CPU LLVM / GPU OpenCL / CPU OpenCL / GPU metal ]

~/.keras/keras.json:
{
		"floatx": "float32",
		"epsilon": 1e-07,
		"backend": "tensorflow",
		"image_data_format": "channels_last"
	}

plaidml 0.3.4, keras 2.0.8, tensorflow 1.9.0, plaidbench 0.3.2, virtualenv 16.0.0, python 3.6.4

iMac, OSX 10.13.4, AMD Rpro 580

**Comment by AlexSmithDeveloper** [comment]
The workaround proposed by @mzeug worked for me.

Trying to test my installation, it was giving me this error.
```
PS C:\WINDOWS\system32> plaidbench keras mobilenet
Running 1024 examples with mobilenet, batch size 1
INFO:plaidml:Opening device "opencl_amd_ellesmere.0"
could not broadcast input array from shape (3,2048) into shape (6144)

```
After mzeug's suggested changes to the __init__.py file...
> for me, the file was located at
> C:\Python37\Lib\site-packages\plaidml\__init__.py
> 
> in both of these methods, "copy_from_ndarray" and "copy_to_ndarray" on lines 1193 and 1197
>
> np.ctypeslib.as_array(self, shape=src.shape)
> changed to 
> np.ctypeslib.as_array(self._base, shape=src.shape)
>
 it seems like the operation was able to be completed.

```
PS C:\WINDOWS\system32> plaidbench keras mobilenet
Running 1024 examples with mobilenet, batch size 1
INFO:plaidml:Opening device "opencl_amd_ellesmere.0"
could not broadcast input array from shape (3,2048) into shape (6144)
Set --print-stacktraces to see the entire traceback
PS C:\WINDOWS\system32> plaidbench keras mobilenet
Running 1024 examples with mobilenet, batch size 1
INFO:plaidml:Opening device "opencl_amd_ellesmere.0"
Downloading data from https://github.com/fchollet/deep-learning-models/releases/download/v0.6/mobilenet_1_0_224_tf.h5
16809984/17225924 [============================>.] - ETA: 0sModel loaded.
Compiling network...
Warming up ...
Main timing
Example finished, elapsed: 4.597696304321289 (compile), 6.579812526702881 (execution), 0.006425598170608282 (execution per example)
Correctness: PASS, max_error: 1.8053706298815086e-05, max_abs_error: 9.760260581970215e-07, fail_ratio: 0.0
```
Windows 10 Pro 64-bit | CPU: AMD Ryzen 1800X | GPU: AMD RX580 8GB | python 3.7.0

I have basically no idea what I'm doing, but I hope this helps.

**Comment by rlong011** [comment]
AlexSmithDeveloper
I to am learning...
I have Ubuntu 16.04.3 | AMD FX6300 |5x Rx 480 | python 3.5

I found the plaidml _init_py file at  /usr/local/lib/python3.5/dist-packages/plaidml/ def 

I made the changes as you and @mzeug suggested:

def copy_from_ndarray(self, src):
        if self._dtype == DType.FLOAT16:
            if src.dtype != 'float16':
                src = src.astype('float16')
            src = src.view(dtype='uint16')
        #dst = np.ctypeslib.as_array(self, shape=src.shape)
        **dst = np.ctypeslib.as_array(self._base, shape=src.shape)**
        np.copyto(dst, src)

    def copy_to_ndarray(self, dst):
        #src = np.ctypeslib.as_array(self, shape=dst.shape)
        **src = np.ctypeslib.as_array(self._base, shape=src.shape)**
        if self._dtype == DType.FLOAT16:
            src = src.view(dtype='float16')
        np.copyto(dst, src)

I must have done something wrong, I gets further but has a problem in the same area

File "/usr/local/lib/python3.5/dist-packages/plaidml/__init__.py", line 1253, in as_ndarray
    view.copy_to_ndarray(self._ndarray)
  File "/usr/local/lib/python3.5/dist-packages/plaidml/__init__.py", line 1199, in copy_to_ndarray
    src = np.ctypeslib.as_array(self._base, shape=src.shape)
UnboundLocalError: local variable 'src' referenced before assignment

As I have little idea what I'm doing yet, any help would be appreciated

**Comment by mzeug** [comment]
@rlong011 
in the copy_to_ndarray you must not change the shape parameter. Sorry i didn't cover this.
`np.ctypeslib.as_array(self._base, shape=dst.shape)`
should fix your propblem

But i have to make clear that this isn't a proper workaround! My understanding of plaidML codebase is zero (working with ctypes is also nothing i've done that often) ... that was just a lucky guess based on the given error message and some documentation reading that somehow worked and produced reasonable results for the data I tested and compared to a normal TF-Backend keras .

:smile:

**Comment by dr-romster** [comment]
Great thanks for your help

In my python 3.6 virtual environment the path to the file was 

_~/plaidml-venv/lib/python3.6/site-packages/plaidml/__init__.py_

(where 'plaid-venv' is the activated virtual environment)

The line numbers to edit were as above in @mzeug 's post [**1193** & **1197**]. 

The `plaidbench` test works fine with both OpenCL and Metal backends. 
Trying to `plaidml.keras.install_backend()` with metal fails, but has no issue with OpenCL. I can live with that for now!

**Comment by rlong011** [comment]
Thanks very much, it was me I just copied from the copy_from_ndarray. It's running  now and appeared to complete all 25 fo 25 epochs

**Comment by tzerrell** [comment]
Thanks @mzeug , that is a good solution for this bug! We're putting together a release that will include this bugfix; in the interim, if you're experiencing this bug you can use @mzeug 's fix (@dr-romster 's post above gives good details about where to make the changes).

**Comment by liebmann** [comment]
**Downgrading numpy to version 1.14.5 fixes the problem!** (MacOS High Sierra 10.13.6)

```
pip uninstall numpy==1.15.1
pip install numpy==1.14.5
```

**Comment by tzerrell** [comment]
This should be fixed in PlaidML 0.3.5; if anyone continues to see this error please let us know.
